src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs

```diff
@@ -2,20 +2,62 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
 namespace System.Globalization
 {
     internal static partial class GlobalizationMode
     {
         private static bool GetInvariantSwitchValue() =>
             GetSwitchValue("System.Globalization.Invariant", "DOTNET_SYSTEM_GLOBALIZATION_INVARIANT");
 
+        private static bool TryGetAppLocalIcuSwitchValue(out ReadOnlySpan<char> version, out ReadOnlySpan<char> icuSuffix)
+        {
+            icuSuffix = default;
+
+            if (!TryGetStringValue("System.Globalization.AppLocalIcu", "DOTNET_SYSTEM_GLOBALIZATION_APPLOCALICU", out string? value))
+            {
+                version = default;
+                return false;
+            }
+
+            // Custom built ICU can have a suffix on the name, i.e: libicuucmyapp.so.67.1
+            // So users would set the runtime switch as: myapp:67.1
+
+            int indexOfSeparator = value.IndexOf(':', StringComparison.Ordinal);
+            if (indexOfSeparator != -1)
+            {
+                ReadOnlySpan<char> valueAsSpan = value.AsSpan();
+                icuSuffix = valueAsSpan.Slice(0, indexOfSeparator);
+
+                if (icuSuffix.Length > 20)
+                {
+                    Environment.FailFast($"The resolved \"{icuSuffix.ToString()}\" suffix from System.Globalization.AppLocalIcu switch has to be < 20 chars long.");
+                }
+
+                version = valueAsSpan.Slice(icuSuffix.Length + 1);
```

Comment:
>version = valueAsSpan.Slice(icuSuffix.Length + 1); [](start = 16, length = 50)

did you try to see what will happen when ':' is the last character in the string?

Reply:
Nope. I'll add validation... should we just assume the version in that case is empty and we try to load an un-versioned ICU?

Reply:
Hm, this raises an interesting question/point: it is possible to build ICU *without* any version or suffix at all. TBH, I think that is probably is more common of a scenario than adding a custom suffix, as it lets you link to the same API surface across versions. Assuming that you are only using public/stable flat C APIs, it gives binary compat.

Reply:
Yes that is reasonable to try to load without version at that time. This is similar if the whole string is empty too.

---
In reply to: [414744435](https://github.com/dotnet/runtime/pull/35383#discussion_r414744435) [](ancestors = 414744435)

Reply:
On Linux, the .so files without version number are development-time libraries (they are just a link to the actual library), but if you don't install the development package for that library, they are not present.

Reply:
@jefgen I believe it is good to be more flexible and allow loading modules with no versions. The bottom line is, if the module exist we'll load it.

Reply:
What format do you suggest for the runtime switch to be when I want to say, don't use any version?

Reply:
Sorry was thinking of on Windows platforms. The DLL names would normally always have the major version appended, though if you wanted you could edit the MSbuild/VS project files to remove it. The actual import lib doesn't have the version. It is possible to build without suffixes appended to the API names inside the module, though the dll would still have it. I'm not sure how common that is though.

Reply:
> The DLL names would normally always have the major version appended

@safern did you count for that? 

@jefgen does the API names can change (if there is a suffix)? looks we need to pay more attention to what you are saying here.

Reply:
I'll try to add some more details/info:
On Linux, you can run the configure with the option  `--disable-renaming`. This will prevent the version suffix from being added the API names.

However the actual .so files would still have a version on them. Symlinks are created for the name without a version that points to the "versioned" .so file.

To give an example:
Here’s what you get if you use the following to build ICU on Linux:
`./runConfigureICU Linux --disable-renaming --prefix=/tmp/icu-67-temp && make all && make install`
```
$ ls -al /tmp/icu-67-temp/lib 
total 34584 
drwxrwxr-x 4 jeff jeff     4096 Apr 24 10:53 ./ 
drwxrwxr-x 7 jeff jeff     4096 Apr 24 10:53 ../ 
drwxrwxr-x 3 jeff jeff     4096 Apr 24 10:53 icu/ 
lrwxrwxrwx 1 jeff jeff       18 Apr 24 10:53 libicudata.so -> libicudata.so.67.1* 
lrwxrwxrwx 1 jeff jeff       18 Apr 24 10:53 libicudata.so.67 -> libicudata.so.67.1* 
-rwxr-xr-x 1 jeff jeff 28398712 Apr 24 10:53 libicudata.so.67.1* 
lrwxrwxrwx 1 jeff jeff       18 Apr 24 10:53 libicui18n.so -> libicui18n.so.67.1* 
lrwxrwxrwx 1 jeff jeff       18 Apr 24 10:53 libicui18n.so.67 -> libicui18n.so.67.1* 
-rwxr-xr-x 1 jeff jeff  4139000 Apr 24 10:53 libicui18n.so.67.1*
```

In this setup you’ll want/need to set the following *before* using the ICU headers:
`#define U_DISABLE_RENAMING 1`

This will omit the version suffix on the API names.
Example: The API `u_strlen_67` would change to just `u_strlen`.

On Windows, you could edit the MSBuild props files to changed/modify `PreprocessorDefinitions` to add `U_DISABLE_RENAMING=1` which would accomplish the same thing. However, the DLLs that are produced would still look like `icuuc67.dll`. (Unless you _also_ edited the MSbuild files to remove it as well).

I don’t know how common it would be to remove the version number from the DLL, but building with either `--disable-renaming` or `U_DISABLE_RENAMING` set is likely to occur for anyone that wants to have a stable API surface for binary compat.

Reply:
Thanks @jefgen -- yeah, we do account for that. The first combination of names we try on the exported symbols is without a version, so this scenario should be already covered in the native code when we try to find the symbol name format:
https://github.com/dotnet/runtime/pull/35383/files#diff-9812d22f4140085b5587ec72c0d99cb0R53
https://github.com/dotnet/runtime/pull/35383/files#diff-9812d22f4140085b5587ec72c0d99cb0R114

Reply:
Nice! Thanks for the pointer @safern.